### ISSUE TITLE
TypeError: Cannot read property 'recalculate' of undefined

### ISSUE THREAD

**Issue by jfirebaugh** [mapbox/mapbox-gl-js#3345]
Seen in current (208d3539d736b0a0e2bdb2e3447ca37ae8319b20) `npm test` output. Although it's not resulting in a test failure, we don't expect error output to be printed:

```
TypeError: Cannot read property 'recalculate' of undefined
    at Object._recalculate (/Users/distiller/mapbox-gl-js/js/style/style.js:4:3301)
    at _render (/Users/distiller/mapbox-gl-js/js/ui/map.js:438:423)
    at runCallback (timers.js:576:20)
    at tryOnImmediate (timers.js:554:5)
    at processImmediate [as _immediateCallback] (timers.js:533:5)
```

**Comment by jfirebaugh** [comment]
Example: https://circleci.com/gh/mapbox/mapbox-gl-js/5285

**Comment by jfirebaugh** [comment]
[This line is throwing](https://github.com/mapbox/mapbox-gl-js/blob/c8cbb9d6b0d541413347f317ae7c181fbf62d110/js/style/style.js#L240-L240) because `this.light` is undefined.

The exception propagates to the [catch block in `Map#_render`](https://github.com/mapbox/mapbox-gl-js/blob/c8cbb9d6b0d541413347f317ae7c181fbf62d110/js/ui/map.js#L1072-L1074) where it's emitted as an `error` event. Nothing is listening to this event so [Evented logs it](https://github.com/mapbox/mapbox-gl-js/blob/24f2135f058372bd7906e3c09148c92c8ffebef0/js/util/evented.js#L91-L91) and then the tests continue.

@lucaswoj Didn't we try to put something in place so that tests will fail in a situation like this?

**Comment by lucaswoj** [comment]
It would appear that we do not listen to the `error` event within `suite_implementation`. We should fix that.

**Comment by jfirebaugh** [comment]
@lucaswoj This is in unit tests, not render tests.

**Comment by lucaswoj** [comment]
Ah. Likewise for unit tests. We do not have a mechanism to throw `error` events.

Is there any reason we can't throw errors within `Evented` if there are no listeners rather than print them?

**Comment by mourner** [comment]
@lucaswoj this might be a good solution. It should in theory also help with stack traces because errors will get thrown before they are propagated through a chain of parents.

**Comment by jfirebaugh** [comment]
That would change the execution path of any existing code that does `fire('error', ...)`. That's probably not what we want.

My preferred fix is to revert #2548. It made debugging more difficult and doesn't seem to accomplish its goal of not causing rendering to stop, as was noted in https://github.com/mapbox/mapbox-gl-js/pull/2548#r62579732.

**Comment by lucaswoj** [comment]
> That would change the execution path of any existing code that does fire('error', ...). That's probably not what we want.

Right. It would be weird if the control flow of `fire('error', ...` was different depending on whether or not listeners are bound.
